Deprecated use of `dyn_cast` to excise

#16011 fixed some warning flag issues that were masking deprecation warnings. We now have a few in a clean build from head.

```
In file included from /home/stella/src/iree/compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:11:
In file included from /home/stella/src/iree/compiler/src/iree/compiler/Dialect/Flow/IR/FlowOps.h:13:
In file included from /home/stella/src/iree/compiler/src/iree/compiler/Dialect/Flow/IR/FlowTypes.h:11:
In file included from /home/stella/src/iree/compiler/src/iree/compiler/Dialect/Util/IR/UtilTypes.h:11:
/home/stella/src/iree/third_party/llvm-project/llvm/include/llvm/ADT/TypeSwitch.h:77:27: warning: 'dyn_cast' is deprecated: Use llvm::dyn_cast<U>() instead [-Wdeprecated-declarations]
    return value.template dyn_cast<CastT>();
                          ^
/home/stella/src/iree/third_party/llvm-project/llvm/include/llvm/ADT/TypeSwitch.h:121:42: note: in instantiation of function template specialization 'llvm::detail::TypeSwitchBase<llvm::TypeSwitch<mlir::AffineExpr, bool>, mlir::AffineExpr>::castValue<mlir::AffineDimExpr, const mlir::AffineExpr &>' requested here
    if (auto caseValue = BaseT::template castValue<CaseT>(this->value))
                                         ^
/home/stella/src/iree/compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:337:29: note: in instantiation of function template specialization 'llvm::TypeSwitch<mlir::AffineExpr, bool>::Case<mlir::AffineDimExpr, (lambda at /home/stella/src/iree/compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:337:49)>' requested here
                           .Case<AffineDimExpr>([&dimsSeen](auto dimExpr) {
                            ^
/home/stella/src/iree/third_party/llvm-project/mlir/include/mlir/IR/AffineExpr.h:289:15: note: 'dyn_cast' has been explicitly marked deprecated here
U AffineExpr::dyn_cast() const {
              ^
In file included from /home/stella/src/iree/compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:11:
In file included from /home/stella/src/iree/compiler/src/iree/compiler/Dialect/Flow/IR/FlowOps.h:13:
In file included from /home/stella/src/iree/compiler/src/iree/compiler/Dialect/Flow/IR/FlowTypes.h:11:
In file included from /home/stella/src/iree/compiler/src/iree/compiler/Dialect/Util/IR/UtilTypes.h:11:
/home/stella/src/iree/third_party/llvm-project/llvm/include/llvm/ADT/TypeSwitch.h:77:27: warning: 'dyn_cast<mlir::AffineDimExpr>' is deprecated: Use llvm::dyn_cast<U>() instead [-Wdeprecated-declarations]
    return value.template dyn_cast<CastT>();
                          ^
/home/stella/src/iree/third_party/llvm-project/mlir/include/mlir/IR/AffineExpr.h:88:5: note: 'dyn_cast<mlir::AffineDimExpr>' has been explicitly marked deprecated here
  [[deprecated("Use llvm::dyn_cast<U>() instead")]] U dyn_cast() const;
    ^
In file included from /home/stella/src/iree/compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:11:
In file included from /home/stella/src/iree/compiler/src/iree/compiler/Dialect/Flow/IR/FlowOps.h:13:
In file included from /home/stella/src/iree/compiler/src/iree/compiler/Dialect/Flow/IR/FlowTypes.h:11:
In file included from /home/stella/src/iree/compiler/src/iree/compiler/Dialect/Util/IR/UtilTypes.h:11:
/home/stella/src/iree/third_party/llvm-project/llvm/include/llvm/ADT/TypeSwitch.h:77:27: warning: 'dyn_cast' is deprecated: Use llvm::dyn_cast<U>() instead [-Wdeprecated-declarations]
    return value.template dyn_cast<CastT>();
                          ^
/home/stella/src/iree/third_party/llvm-project/llvm/include/llvm/ADT/TypeSwitch.h:121:42: note: in instantiation of function template specialization 'llvm::detail::TypeSwitchBase<llvm::TypeSwitch<mlir::AffineExpr, bool>, mlir::AffineExpr>::castValue<mlir::AffineConstantExpr, const mlir::AffineExpr &>' requested here
    if (auto caseValue = BaseT::template castValue<CaseT>(this->value))
                                         ^
/home/stella/src/iree/compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:343:29: note: in instantiation of function template specialization 'llvm::TypeSwitch<mlir::AffineExpr, bool>::Case<mlir::AffineConstantExpr, (lambda at /home/stella/src/iree/compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:343:54)>' requested here
                           .Case<AffineConstantExpr>([](auto constExpr) {
                            ^
/home/stella/src/iree/third_party/llvm-project/mlir/include/mlir/IR/AffineExpr.h:289:15: note: 'dyn_cast' has been explicitly marked deprecated here
U AffineExpr::dyn_cast() const {
              ^
In file included from /home/stella/src/iree/compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:11:
In file included from /home/stella/src/iree/compiler/src/iree/compiler/Dialect/Flow/IR/FlowOps.h:13:
In file included from /home/stella/src/iree/compiler/src/iree/compiler/Dialect/Flow/IR/FlowTypes.h:11:
In file included from /home/stella/src/iree/compiler/src/iree/compiler/Dialect/Util/IR/UtilTypes.h:11:
/home/stella/src/iree/third_party/llvm-project/llvm/include/llvm/ADT/TypeSwitch.h:77:27: warning: 'dyn_cast<mlir::AffineConstantExpr>' is deprecated: Use llvm::dyn_cast<U>() instead [-Wdeprecated-declarations]
    return value.template dyn_cast<CastT>();
                          ^
/home/stella/src/iree/third_party/llvm-project/mlir/include/mlir/IR/AffineExpr.h:88:5: note: 'dyn_cast<mlir::AffineConstantExpr>' has been explicitly marked deprecated here
  [[deprecated("Use llvm::dyn_cast<U>() instead")]] U dyn_cast() const;
    ^
4 warnings generated.

```

--- a/compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp
+++ b/compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp
@@ -317,20 +317,19 @@ static bool isIdentityMapWithZeros(AffineMap map) {
   if (map.isEmpty())
     return false;
   unsigned dimsSeen = 0;
-  for (auto result : map.getResults()) {
-    bool isValidExpr = TypeSwitch<AffineExpr, bool>(result)
-                           .Case<AffineDimExpr>([&dimsSeen](auto dimExpr) {
-                             if (dimExpr.getPosition() != dimsSeen)
-                               return false;
-                             dimsSeen++;
-                             return true;
-                           })
-                           .Case<AffineConstantExpr>([](auto constExpr) {
-                             return constExpr.getValue() == 0;
-                           })
-                           .Default([](AffineExpr) { return false; });
-    if (!isValidExpr)
+  for (AffineExpr result : map.getResults()) {
+    if (auto dimExpr = dyn_cast<AffineDimExpr>(result)) {
+      if (dimExpr.getPosition() != dimsSeen) {
+        return false;
+      }
+      dimsSeen++;
+    } else if (auto constExpr = dyn_cast<AffineConstantExpr>(result)) {
+      if (constExpr.getValue() != 0) {
+        return false;
+      }
+    } else {
       return false;
+    }
   }
   return dimsSeen == map.getNumDims();
 }

